File: data/derived_cot/rq3_prompting/results/translation/r1/py_to_cpp/code_output/KappaCalculator.cpp
#include <vector>
#include <cstddef>

class KappaCalculator {
public:
    static double kappa(const std::vector<std::vector<int>>& testData, int k) {
        // Convert to double vectors for arithmetic
        std::vector<std::vector<double>> data(k, std::vector<double>(k));
        double sum = 0.0;
        for (int i = 0; i < k; ++i) {
            for (int j = 0; j < k; ++j) {
                data[i][j] = static_cast<double>(testData[i][j]);
                sum += data[i][j];
            }
        }

        // Diagonal sum (P0)
        double P0 = 0.0;
        for (int i = 0; i < k; ++i) {
            P0 += data[i][i];
        }
        double P0_ratio = P0 / sum;

        // Row sums and column sums
        std::vector<double> rowSums(k, 0.0);
        std::vector<double> colSums(k, 0.0);
        for (int i = 0; i < k; ++i) {
            for (int j = 0; j < k; ++j) {
                rowSums[i] += data[i][j];
                colSums[j] += data[i][j];
            }
        }

        // Pe = (colSums dot rowSums) / (sum * sum)
        double Pe = 0.0;
        for (int i = 0; i < k; ++i) {
            Pe += colSums[i] * rowSums[i];
        }
        Pe /= (sum * sum);

        double kappa = (P0_ratio - Pe) / (1.0 - Pe);
        return kappa;
    }

    static double fleiss_kappa(const std::vector<std::vector<int>>& testData, int N, int k, int n) {
        // Convert to double matrix
        std::vector<std::vector<double>> data(N, std::vector<double>(k));
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < k; ++j) {
                data[i][j] = static_cast<double>(testData[i][j]);
            }
        }

        double totalSum = 0.0;
        double P0 = 0.0;
        std::vector<double> colSums(k, 0.0);

        for (int i = 0; i < N; ++i) {
            double temp = 0.0;
            for (int j = 0; j < k; ++j) {
                double val = data[i][j];
                totalSum += val;
                temp += val * val;
                colSums[j] += val;
            }
            temp -= static_cast<double>(n);
            temp /= static_cast<double>((n - 1) * n);
            P0 += temp;
        }
        P0 /= static_cast<double>(N);

        double Pe = 0.0;
        for (int j = 0; j < k; ++j) {
            double pj = colSums[j] / totalSum;
            Pe += pj * pj;
        }

        double ans = (P0 - Pe) / (1.0 - Pe);
        return ans;
    }
};

File: data/derived_cot/rq1_traces/translation/qwen/java_to_py/code_output/KappaCalculator.py
def kappa(test_data, k):
    data_mat = test_data
    total_observations = 0
    P0 = 0.0
    for i in range(k):
        P0 += data_mat[i][i]
    row_sums = [0] * k
    col_sums = [0] * k
    for i in range(k):
        for j in range(k):
            row_sums[i] += data_mat[i][j]
            col_sums[j] += data_mat[i][j]
            total_observations += data_mat[i][j]
    Pe = 0.0
    for i in range(k):
        Pe += col_sums[i] * row_sums[i]
    Pe = Pe / (total_observations ** 2)
    P0 = P0 / total_observations
    return (P0 - Pe) / (1 - Pe)

def fleiss_kappa(test_data, N, k, n):
    P = [0.0] * N
    total_sum = 0.0
    for i in range(N):
        row_total = 0
        temp = 0.0
        for j in range(k):
            row_total += test_data[i][j]
            temp += test_data[i][j] * test_data[i][j]
        temp -= n
        temp /= (n - 1) * n
        P[i] = temp
    P0 = sum(P) / N
    pj = [0.0] * k
    for j in range(k):
        for i in range(N):
            pj[j] += test_data[i][j]
    for j in range(k):
        pj[j] /= total_sum
    Pe = sum(p * p for p in pj)
    return (P0 - Pe) / (1 - Pe)